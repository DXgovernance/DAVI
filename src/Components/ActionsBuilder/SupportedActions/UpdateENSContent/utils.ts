import contenthash from 'content-hash';
import { MAINNET_ID, LOCALHOST_ID, GOERLI_ID } from 'utils';
import { utils } from 'ethers';

const DEFAULT_NAMEHASH =
  '0x0000000000000000000000000000000000000000000000000000000000000000';

export const convertToNameHash = (name: string) => {
  if (!name) {
    return DEFAULT_NAMEHASH;
  }
  return utils.namehash(name);
};

export const convertToContentHash = (ipfsHash: string) => {
  if (!ipfsHash) {
    return '';
  }
  return contenthash.fromIpfs(ipfsHash);
};

export const convertToIpfsHash = (contentHash: string) => {
  if (!contentHash) return null;
  return contenthash.decode(contentHash);
};

export const getIpfsUrl = (ipfsHash: string) => {
  const ipfsRoot = 'ipfs://';
  return `${ipfsRoot}${ipfsHash}`;
};

export const isValidChainId = (chainId: number) => {
  if (!chainId || chainId === LOCALHOST_ID) {
    return MAINNET_ID;
  } else {
    return chainId;
  }
};

export const isAvailableOnENS = (chainId: number) => {
  const validChainId = isValidChainId(chainId);
  const ensNetworks = [MAINNET_ID, GOERLI_ID];
  return ensNetworks.includes(validChainId);
};
