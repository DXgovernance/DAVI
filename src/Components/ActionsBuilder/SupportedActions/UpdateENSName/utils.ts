import namehash from 'eth-ens-namehash';
import contenthash from 'content-hash';
import { useState, useEffect } from 'react';

export const convertToNameHash = (name: string) => namehash.hash(name);

export const convertToContentHash = (ipfsHash: string) =>
  contenthash.fromIpfs(ipfsHash);

export const convertToIPFSHash = (contentHash: string) => {
  if (!contentHash) return null;
  return contenthash.decode(contentHash);
};

export const useDebounce = (value: string, delay: number) => {
  const [debouncedValue, setDebouncedValue] = useState(value);

  useEffect(() => {
    const handler = setTimeout(() => {
      setDebouncedValue(value);
    }, delay);

    return () => {
      clearTimeout(handler);
    };
  }, [value, delay]);

  return debouncedValue;
};
