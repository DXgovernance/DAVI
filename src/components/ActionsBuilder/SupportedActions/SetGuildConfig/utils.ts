import { FIELDS } from './constants';
import { BigNumber } from 'ethers';
import { SetGuildConfigFields, ControlField } from './types';

export const bn = (n: string | number | BigNumber) => BigNumber.from(n);

type UpdatedValuesReturn = {
  [key in ControlField]: {
    newValue: BigNumber;
    currentValue: BigNumber;
  };
};
export const getUpdatedValues = (
  current: SetGuildConfigFields,
  modifyed: SetGuildConfigFields
): UpdatedValuesReturn | {} => {
  if (!current || !modifyed) return {};
  return Object.keys(current).reduce((acc, key) => {
    if (!FIELDS.map(f => f.name).includes(key)) return acc;
    const currentValue = bn(current[key] || 0);
    const newValue = bn(modifyed[key] || 0);

    if (!currentValue.eq(newValue)) {
      return {
        ...acc,
        [key]: { newValue, currentValue },
      };
    }
    return acc;
  }, {});
};

export const pickUpdatedOrDefaultValue = (
  current: BigNumber,
  modifyed: BigNumber
): BigNumber => {
  if (BigNumber.isBigNumber(current)) {
    if (BigNumber.isBigNumber(modifyed) && !bn(modifyed).eq(current)) {
      return bn(modifyed);
    }
    return bn(current);
  } else if (BigNumber.isBigNumber(modifyed)) {
    return bn(modifyed);
  }
  return bn(0);
};
