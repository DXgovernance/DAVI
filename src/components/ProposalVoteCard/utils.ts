import { ConfirmVoteProposalProps, VoteOnProposalProps } from './types';

export const voteOnProposal = ({
  votingPowerAtProposalSnapshot,
  votingPowerAtProposalCurrentSnapshot,
}: VoteOnProposalProps) => {
  const hasNoVotingPower =
    votingPowerAtProposalSnapshot &&
    Number(votingPowerAtProposalSnapshot?.toString()) <= 0;

  const hasVotingPowerAtCurrentSnapshot =
    votingPowerAtProposalCurrentSnapshot &&
    Number(votingPowerAtProposalCurrentSnapshot?.toString()) > 0;
  return {
    hasNoVotingPower,
    hasVotingPowerAtCurrentSnapshot,
  };
};

export const confirmVoteProposal = ({
  proposal,
  contract,
  selectedOption,
  userVotingPower,
  createTransaction,
  cb,
}: ConfirmVoteProposalProps) => {
  createTransaction(
    `Vote on proposal ${proposal?.title}`,
    async () => contract.setVote(proposal?.id, selectedOption, userVotingPower),
    true,
    cb
  );
};

export const getOptionLabel = ({ metadata, optionKey, t }) => {
  const metadataLabel = metadata?.voteOptions?.[optionKey];
  return metadataLabel
    ? metadataLabel
    : Number(optionKey) === 0
    ? t('against', { defaultValue: 'Against' })
    : t('option', { optionKey });
};
