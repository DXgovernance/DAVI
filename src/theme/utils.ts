import { css } from 'styled-components';

export const SUPPORTED_THEMES = {
  DARK: 'DARK',
  LIGHT: 'LIGHT',
};

const MEDIA_WIDTHS = {
  upToSmall: 600,
  upToMedium: 960,
  upToLarge: 1280,
};

export const mediaWidthTemplates = Object.keys(MEDIA_WIDTHS).reduce(
  (accumulator, size) => {
    accumulator[size] = (...args: any[]) => css`
      @media (max-width: ${MEDIA_WIDTHS[size]}px) {
        ${
          // @ts-ignore
          css(...args)
        }
      }
    `;
    return accumulator;
  },
  {}
);
