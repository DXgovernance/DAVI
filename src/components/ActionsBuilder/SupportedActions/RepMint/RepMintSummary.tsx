import { Avatar } from 'components/Avatar';
import useENSAvatar from 'hooks/Guilds/ens/useENSAvatar';
import { FiArrowRight } from 'react-icons/fi';
import { MAINNET_ID, shortenAddress } from 'utils';
import { ActionViewProps } from '..';
import { Segment } from '../common/infoLine';
import { ReactComponent as Mint } from 'assets/images/mint.svg';
import useBigNumberToNumber from 'hooks/Guilds/conversions/useBigNumberToNumber';
import { useTotalSupply } from 'Modules/Guilds/Hooks//useTotalSupply';
import { useTokenData } from 'Modules/Guilds/Hooks/useTokenData';
import { useTranslation } from 'react-i18next';
import { StyledMintIcon } from './styles';

const RepMintInfoLine: React.FC<ActionViewProps> = ({ decodedCall }) => {
  const { t } = useTranslation();

  const { data } = useTotalSupply({ decodedCall });
  const { tokenData } = useTokenData();

  const totalSupply = useBigNumberToNumber(tokenData?.totalSupply, 18);

  const { ensName, imageUrl } = useENSAvatar(data?.toAddress, MAINNET_ID);

  const roundedRepAmount = useBigNumberToNumber(data?.amount, 16, 3);
  const roundedRepPercent = roundedRepAmount / totalSupply;

  return (
    <>
      <Segment>
        <StyledMintIcon src={Mint} />
      </Segment>
      <Segment>
        {t('mint')} {roundedRepPercent} %
      </Segment>
      <Segment>
        <FiArrowRight />
      </Segment>
      <Segment>
        <Avatar defaultSeed={data?.toAddress} src={imageUrl} size={24} />
      </Segment>
      <Segment>{ensName || shortenAddress(data?.toAddress)}</Segment>
    </>
  );
};

export default RepMintInfoLine;
