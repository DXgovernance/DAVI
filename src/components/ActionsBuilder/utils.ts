import { DecodedAction } from './types';
import { Permission } from './types';
import Web3 from 'web3';
import { NULL_SIGNATURE } from 'utils';
import { SupportedAction } from './types';

const web3 = new Web3();

export const getPermissionArgs = (
  decodedActions: DecodedAction[]
): Permission[] => {
  const permissions = decodedActions.map(decodedAction => {
    // Native transfer of ETH does not have a function
    if (
      decodedAction?.decodedCall?.callType ===
        SupportedAction.NATIVE_TRANSFER ||
      decodedAction?.decodedCall?.callType === SupportedAction.RAW_TRANSACTION
    ) {
      return {
        from: decodedAction?.decodedCall?.from,
        to: decodedAction?.decodedCall?.to,
        callType: decodedAction?.decodedCall?.callType,
        functionSignature: NULL_SIGNATURE,
      };
    }
    const functionInputs = decodedAction?.decodedCall?.function?.inputs?.map(
      input => input.type
    );
    const functionName = decodedAction?.decodedCall?.function?.name;
    const functionSignature = web3.eth.abi.encodeFunctionSignature(
      `${functionName}(${functionInputs.join(',')})`
    );
    return {
      from: decodedAction?.decodedCall?.from,
      to: decodedAction?.decodedCall?.to,
      callType: decodedAction?.decodedCall?.callType,
      functionSignature,
    };
  });
  return permissions;
};
