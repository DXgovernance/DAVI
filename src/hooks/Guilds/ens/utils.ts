// import { isSupportedChainId } from 'components/ActionsBuilder/SupportedActions/UpdateENSContent/utils';
import { MAINNET_ID } from 'utils';

export const isAvailableOnENS = (chainId: number) => {
  const ensNetworks = [MAINNET_ID];
  return ensNetworks.includes(chainId);
};
