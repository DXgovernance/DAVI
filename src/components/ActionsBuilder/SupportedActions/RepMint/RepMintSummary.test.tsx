import { BigNumber } from 'ethers';
import { render } from 'utils/tests';
import RepMintSummary from './RepMintSummary';
import {
  repMintDecodedCallMock,
  repMintEmptyDecodedCallMock,
} from './fixtures';

const mockBigNumber = BigNumber.from(100000000);

jest.mock('hooks/Guilds/ether-swr/ens/useENSAvatar', () => ({
  __esModule: true,
  default: () => ({
    avatarUri: 'test',
    imageUrl: 'test',
    ensName: 'test.eth',
  }),
}));

jest.mock('hooks/Guilds/guild/useTokenData', () => ({
  useTokenData: () => ({
    tokenData: {
      symbol: 'REP',
      decimals: 18,
      name: 'Reputation',
      address: '0x0000000000000000000000000000000000000000',
      totalSupply: mockBigNumber,
    },
  }),
}));

describe('RepMintSummary', () => {
  it('Should match snapshot', () => {
    const { container } = render(
      <RepMintSummary decodedCall={repMintDecodedCallMock} />
    );
    expect(container).toMatchSnapshot();
  });
  it('Should match snapshot in compact mode', () => {
    const { container } = render(
      <RepMintSummary decodedCall={repMintDecodedCallMock} compact={true} />
    );
    expect(container).toMatchSnapshot();
  });
  it('Should match snapshot without data', () => {
    const { container } = render(<RepMintSummary decodedCall={null} />);
    expect(container).toMatchSnapshot();
  });

  it('Should match snapshot without data in compact mode', () => {
    const { container } = render(
      <RepMintSummary decodedCall={null} compact={true} />
    );
    expect(container).toMatchSnapshot();
  });
  it('Should match snapshot with default values', () => {
    const { container } = render(
      <RepMintSummary decodedCall={repMintEmptyDecodedCallMock} />
    );
    expect(container).toMatchSnapshot();
  });
  it('Should match snapshot with default values in compact mode', () => {
    const { container } = render(
      <RepMintSummary
        decodedCall={repMintEmptyDecodedCallMock}
        compact={true}
      />
    );
    expect(container).toMatchSnapshot();
  });
});
