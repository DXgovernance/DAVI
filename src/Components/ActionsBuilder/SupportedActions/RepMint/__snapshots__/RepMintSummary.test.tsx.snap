// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RepMintSummary Should match snapshot 1`] = `
.c3 {
  height: 1.6rem;
  width: 1.6rem;
}

.c4 {
  border-radius: 50%;
  height: 24px;
  width: 24px;
}

.c0 {
  margin-right: 0.5rem;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c1 {
  margin: 0 8px;
  width: undefinedpx;
  height: undefinedpx;
}

.c1 svg path {
  fill: currentColor;
}

.c2 {
  margin: 0;
}

<div>
   
  <span
    class="c0"
  >
    <div
      class="c1 c2"
    >
      <svg>
        mint.svg
      </svg>
    </div>
  </span>
  <span
    class="c0"
  >
    mint
     
    NaN
     %
  </span>
  <span
    class="c0"
  >
    <svg
      fill="none"
      height="1em"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="2"
      viewBox="0 0 24 24"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <line
        x1="5"
        x2="19"
        y1="12"
        y2="12"
      />
      <polyline
        points="12 5 19 12 12 19"
      />
    </svg>
  </span>
  <span
    class="c0"
  >
    <img
      alt="Avatar"
      class="c3 c4"
      data-testid="avatar"
      size="24"
      src="test"
    />
  </span>
  <span
    class="c0"
  >
    test.eth
  </span>
   
</div>
`;

exports[`RepMintSummary Should match snapshot in compact mode 1`] = `
.c3 {
  height: 1.6rem;
  width: 1.6rem;
}

.c4 {
  border-radius: 50%;
  height: 24px;
  width: 24px;
}

.c0 {
  margin-right: 0.5rem;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c1 {
  margin: 0 8px;
  width: undefinedpx;
  height: undefinedpx;
}

.c1 svg path {
  fill: currentColor;
}

.c2 {
  margin: 0;
}

<div>
   
  <span
    class="c0"
  >
    <div
      class="c1 c2"
    >
      <svg>
        mint.svg
      </svg>
    </div>
  </span>
  <span
    class="c0"
  >
    mint
     
    NaN
     %
  </span>
  <span
    class="c0"
  >
    <svg
      fill="none"
      height="1em"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="2"
      viewBox="0 0 24 24"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <line
        x1="5"
        x2="19"
        y1="12"
        y2="12"
      />
      <polyline
        points="12 5 19 12 12 19"
      />
    </svg>
  </span>
  <span
    class="c0"
  >
    <img
      alt="Avatar"
      class="c3 c4"
      data-testid="avatar"
      size="24"
      src="test"
    />
  </span>
  <span
    class="c0"
  >
    test.eth
  </span>
   
</div>
`;

exports[`RepMintSummary Should match snapshot with default values 1`] = `
.c3 {
  height: 1.6rem;
  width: 1.6rem;
}

.c4 {
  border-radius: 50%;
  height: 24px;
  width: 24px;
}

.c0 {
  margin-right: 0.5rem;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c1 {
  margin: 0 8px;
  width: undefinedpx;
  height: undefinedpx;
}

.c1 svg path {
  fill: currentColor;
}

.c2 {
  margin: 0;
}

<div>
   
  <span
    class="c0"
  >
    <div
      class="c1 c2"
    >
      <svg>
        mint.svg
      </svg>
    </div>
  </span>
  <span
    class="c0"
  >
    mint
     
    NaN
     %
  </span>
  <span
    class="c0"
  >
    <svg
      fill="none"
      height="1em"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="2"
      viewBox="0 0 24 24"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <line
        x1="5"
        x2="19"
        y1="12"
        y2="12"
      />
      <polyline
        points="12 5 19 12 12 19"
      />
    </svg>
  </span>
  <span
    class="c0"
  >
    <img
      alt="Avatar"
      class="c3 c4"
      data-testid="avatar"
      size="24"
      src="test"
    />
  </span>
  <span
    class="c0"
  >
    test.eth
  </span>
   
</div>
`;

exports[`RepMintSummary Should match snapshot with default values in compact mode 1`] = `
.c3 {
  height: 1.6rem;
  width: 1.6rem;
}

.c4 {
  border-radius: 50%;
  height: 24px;
  width: 24px;
}

.c0 {
  margin-right: 0.5rem;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c1 {
  margin: 0 8px;
  width: undefinedpx;
  height: undefinedpx;
}

.c1 svg path {
  fill: currentColor;
}

.c2 {
  margin: 0;
}

<div>
   
  <span
    class="c0"
  >
    <div
      class="c1 c2"
    >
      <svg>
        mint.svg
      </svg>
    </div>
  </span>
  <span
    class="c0"
  >
    mint
     
    NaN
     %
  </span>
  <span
    class="c0"
  >
    <svg
      fill="none"
      height="1em"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="2"
      viewBox="0 0 24 24"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <line
        x1="5"
        x2="19"
        y1="12"
        y2="12"
      />
      <polyline
        points="12 5 19 12 12 19"
      />
    </svg>
  </span>
  <span
    class="c0"
  >
    <img
      alt="Avatar"
      class="c3 c4"
      data-testid="avatar"
      size="24"
      src="test"
    />
  </span>
  <span
    class="c0"
  >
    test.eth
  </span>
   
</div>
`;

exports[`RepMintSummary Should match snapshot without data 1`] = `
.c3 {
  height: 1.6rem;
  width: 1.6rem;
}

.c4 {
  border-radius: 50%;
  height: 24px;
  width: 24px;
}

.c0 {
  margin-right: 0.5rem;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c1 {
  margin: 0 8px;
  width: undefinedpx;
  height: undefinedpx;
}

.c1 svg path {
  fill: currentColor;
}

.c2 {
  margin: 0;
}

<div>
   
  <span
    class="c0"
  >
    <div
      class="c1 c2"
    >
      <svg>
        mint.svg
      </svg>
    </div>
  </span>
  <span
    class="c0"
  >
    mint
     
    NaN
     %
  </span>
  <span
    class="c0"
  >
    <svg
      fill="none"
      height="1em"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="2"
      viewBox="0 0 24 24"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <line
        x1="5"
        x2="19"
        y1="12"
        y2="12"
      />
      <polyline
        points="12 5 19 12 12 19"
      />
    </svg>
  </span>
  <span
    class="c0"
  >
    <img
      alt="Avatar"
      class="c3 c4"
      data-testid="avatar"
      size="24"
      src="test"
    />
  </span>
  <span
    class="c0"
  >
    test.eth
  </span>
   
</div>
`;

exports[`RepMintSummary Should match snapshot without data in compact mode 1`] = `
.c3 {
  height: 1.6rem;
  width: 1.6rem;
}

.c4 {
  border-radius: 50%;
  height: 24px;
  width: 24px;
}

.c0 {
  margin-right: 0.5rem;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c1 {
  margin: 0 8px;
  width: undefinedpx;
  height: undefinedpx;
}

.c1 svg path {
  fill: currentColor;
}

.c2 {
  margin: 0;
}

<div>
   
  <span
    class="c0"
  >
    <div
      class="c1 c2"
    >
      <svg>
        mint.svg
      </svg>
    </div>
  </span>
  <span
    class="c0"
  >
    mint
     
    NaN
     %
  </span>
  <span
    class="c0"
  >
    <svg
      fill="none"
      height="1em"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="2"
      viewBox="0 0 24 24"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <line
        x1="5"
        x2="19"
        y1="12"
        y2="12"
      />
      <polyline
        points="12 5 19 12 12 19"
      />
    </svg>
  </span>
  <span
    class="c0"
  >
    <img
      alt="Avatar"
      class="c3 c4"
      data-testid="avatar"
      size="24"
      src="test"
    />
  </span>
  <span
    class="c0"
  >
    test.eth
  </span>
   
</div>
`;
